rope/base/oi/type_hinting/evaluate.py

```diff
@@ -0,0 +1,393 @@
+# Based on super lightweight Simple Top-Down Parser from http://effbot.org/zone/simple-top-down-parsing.htm
+# and https://bitbucket.org/emacsway/sqlbuilder/src/default/sqlbuilder/smartsql/contrib/evaluate.py
+import re
+from rope.base.utils import pycompat
+from rope.base.oi.type_hinting import utils
+from rope.base import utils as base_utils
+
+
+class SymbolBase(object):
+
+    name = None  # node/token type name
+
+    def __init__(self):
+        self.value = None  # used by name and literals
+        self.first = None
+        self.second = None
+        self.third = None  # used by tree nodes
+
+    def nud(self, parser):
+        raise SyntaxError(
+            "Syntax error (%r)." % self.name
+        )
+
+    def led(self, left, parser):
+        raise SyntaxError(
+            "Unknown operator (%r)." % self.name
+        )
+
+    def evaluate(self, pyobject):
+        raise NotImplementedError((self.name, vars(self)))
```

Comment:
yes, no problem. But state of object is more important in this case. It's not principled.